include/secp256k1_frost.h

```diff
@@ -17,6 +17,84 @@ extern "C" {
  * (https://crysp.uwaterloo.ca/software/frost/).
  */
 
+/** Opaque data structures
+ *
+ *  The exact representation of data inside is implementation defined and not
+ *  guaranteed to be portable between different platforms or versions. If you
+ *  need to convert to a format suitable for storage, transmission, or
+ *  comparison, use the corresponding serialization and parsing functions.
+ */
+
+/** Opaque data structure that holds a signer's _secret_ share.
+ *
+ *  Guaranteed to be 36 bytes in size. Serialized and parsed with
+ *  `frost_share_serialize` and `frost_share_parse`.
+ */
+typedef struct {
+    unsigned char data[36];
+} secp256k1_frost_share;
+
+/** Serialize a FROST share
+ *
+ *  Returns: 1 when the share could be serialized, 0 otherwise
+ *  Args:    ctx: a secp256k1 context object
+ *  Out:   out32: pointer to a 32-byte array to store the serialized share
+ *  In:    share: pointer to the share
+ */
+SECP256K1_API int secp256k1_frost_share_serialize(
+    const secp256k1_context *ctx,
+    unsigned char *out32,
+    const secp256k1_frost_share *share
+) SECP256K1_ARG_NONNULL(1) SECP256K1_ARG_NONNULL(2) SECP256K1_ARG_NONNULL(3);
+
+/** Parse a FROST share.
+ *
+ *  Returns: 1 when the share could be parsed, 0 otherwise.
+ *  Args:    ctx: a secp256k1 context object
+ *  Out:   share: pointer to a share object
+ *  In:     in32: pointer to the 32-byte share to be parsed
+ */
+SECP256K1_API int secp256k1_frost_share_parse(
+    const secp256k1_context *ctx,
+    secp256k1_frost_share *share,
+    const unsigned char *in32
+) SECP256K1_ARG_NONNULL(1) SECP256K1_ARG_NONNULL(2) SECP256K1_ARG_NONNULL(3);
+
+/** Creates key generation shares
+ *
+ *  To generate a key, a trusted dealer generates a share for each other
+ *  participant.
+ *
+ *  Each participant _must_ have a secure channel with the trusted dealer with
+ *  which they can transmit shares to each other.
+ *
+ *  A new seed32 _must_ be used for each key generation session. The trusted
+ *  dealer must NOT REUSE their respective seed32 again for subsequent key
+ *  generation sessions. If a trusted dealer fails to complete this session or
+ *  start a new session to generate a new key, they must NOT REUSE their
+ *  respective seed32 again, but instead generate a new one. It is recommended
+ *  to always choose seed32 uniformly at random to avoid their reuse.
+ *
+ *  Returns: 0 if the arguments are invalid, 1 otherwise
+ *  Args:            ctx: pointer to a context object
+ *  Out:          shares: pointer to the key generation shares
+ *             pubshares: pointer to the public verification shares
+ *                    pk: pointer to the x-only public key
```

Comment:
Should this be an x-only key? I'm not entirely sure, I would need to think about it.

Reply:
The function [conditionally inverts the shares](https://github.com/BlockstreamResearch/secp256k1-zkp/pull/278/files/a6af50e9a4dc70768108278438c391900038d980..cf31f983ff6677e8f36aec542d03f585eaeb13be#diff-65b267bfb41db84f9bf003641ec64ef27687583f9692adb7869f561f14da6d08R134) to guarantee that the private key will match the x-only key.

Since the primary use-case will be on-chain signing, I think it's a useful simplification to take care of the negation during share generation and makes it easier to reason about in [secp256k1_frost_pubkey_get](https://github.com/BlockstreamResearch/secp256k1-zkp/pull/278/commits/772a0386c1a5f250c974efa6615972cd043f7d7f#diff-65b267bfb41db84f9bf003641ec64ef27687583f9692adb7869f561f14da6d08R187) and [secp256k1_frost_pubkey_tweak](https://github.com/BlockstreamResearch/secp256k1-zkp/pull/278/commits/772a0386c1a5f250c974efa6615972cd043f7d7f#diff-65b267bfb41db84f9bf003641ec64ef27687583f9692adb7869f561f14da6d08R203).

Reply:
Upon further reflection, I'm realizing it's not a useful simplification because in practice the output of the DKG will typically be further tweaked by a BIP32 tweak and then a taproot tweak, rather than used directly.

So I now think it would be better to have the DKG output a 33-byte key rather than an x-only key.

Reply:
I've taken a stab at implementing this change, however, it requires tracking more state during signing.

Currently, `nonce_process` takes an xonly `agg_pk` and optionally a `tweak_cache` and returns a `session`. `partial_sign` takes a `session` and optionally a `tweak_cache`.

`partial_sign` only needs to handle negation logic if a `tweak_cache` is present because the DKG outputs an x-only key. If we change the DKG to output a 33-byte key, then `nonce_process` will need to take a 33-byte `agg_pk` and it will need to save a parity bit, serialized as a byte, in the `session`, so that `partial_sign` can handle the negation logic for when a `tweak_cache` is not present.

So while I don't think there's any reason for the protocol to output a 33-byte key, it seems useful to have the implementation do so, because it avoids a serialized byte in the `session` and I don't see a downside to doing so.

Reply:
I need to compare this approach to the one taken in the signing BIP. The BIP proposes that the pubshares be saved to the `session` and the 33 byte public key is derived from the pubshares.

Reply:
https://github.com/siv2r/bip-frost-signing/issues/12

Reply:
Perhaps the best reason to not output an xonly key from the DKG is that we don't want to encourage FROST DKG outputs being used directly on-chain without an unspendable script path because unlike MuSig, FROST does not randomize the group public key: https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#cite_note-23

Reply:
The signing BIP now [uses signer public shares](https://github.com/siv2r/bip-frost-signing/commit/3f23cf9f2f2177c00048782c93b90caa84b61ca8) to initialize the tweak context (name to be updated) rather than the group public key. Hence, it no longer depends on whether the group pubkey (produced by keygen) is plain or x-only.

Reply:
An alternative design choice here is for `trusted_gen` to always output `tweak_cache` and optionally `group_pk`. This would be similar to how `musig_pubkey_agg` [outputs `keyagg_cache` and optionally `agg_pk`](https://github.com/BlockstreamResearch/secp256k1-zkp/blob/master/include/secp256k1_musig.h#L203-L207).